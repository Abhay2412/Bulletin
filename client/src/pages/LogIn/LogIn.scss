@import "../../styles/colors";

.login-container {
  height: 100%;
  width: 100%;
  .login {
    height: 100%;
    width: 100%;
    margin: 0;

    &-app-description {
      color: $color-white;
      text-align: center;
      margin-top: -3px;
    }

    &-left-container {
      background-color: $color-geyser;
      height: 100%;
      padding: 50px 0px;
      width: 100%;
    }

    &-app-name {
      color: $color-lynch;
      margin-top: -5px;
    }

    &-signup-redirect-text {
      color: $color-lynch;
    }

    &-signup-redirect-button {
      background-color: $color-pink;
      color: $color-white;
      margin-top: -10px;
      width: 40%;
      min-width: 150px;
    }

    &-signup-redirect-button:hover {
      background-color: $color-lynch;
    }

    &-text-field {
      margin-top: -10px;
      width: 100%;
      min-width: 200px;
    }

    &-title {
      color: $color-brown;
    }

    &-button {
      background-color: $color-pink;
      color: $color-brown;
      width: 40%;
      min-width: 150px;
      margin-top: 50px;
    }

    &-inner-grid {
      width: 100%;
    }

    &-inner-grid-item {
      width: 60%;
      min-width: 200px;
    }
  }
}
