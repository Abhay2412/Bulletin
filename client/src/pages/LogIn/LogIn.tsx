import axios from "axios";
import axiosInstance from "../../axios";
import React, { useState } from "react";
import { Grid, Button, TextField } from "@material-ui/core";
import Logo from "../../assets/logo.svg";
import { signInEndPoint } from "../../authEndPoints";
import { useTypedDispatch } from "../../hooks/ReduxHooks";
import { userLoggedIn } from "../../actions/UserActions/UserActionCreator";
import { useNavigate } from "react-router-dom";
import SpinnerButton from "../../components/SpinnerButton";
import "./LogIn.scss";
import { format } from "date-fns";

interface LogInErrors {
  email: string;
  password: string;
}

const LogIn: React.FC = () => {
  const dispatch = useTypedDispatch();
  const navigate = useNavigate();
  const [loginLoading, setLoginLoading] = useState(false);
  const [email, setEmail] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [errors, setErrors] = useState<LogInErrors>({
    email: "",
    password: "",
  });

  const validateData = () => {
    errors.email = "";
    errors.password = "";
    let errorsExits = false;
    if (!email) {
      errors.email = "Please enter an email.";
      errorsExits = true;
    } else if (
      !/^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(
        email
      )
    ) {
      errors.email = "Please enter a valid email.";
      errorsExits = true;
    }

    if (!password) {
      errors.password = "Please enter a password.";
      errorsExits = true;
    }

    setErrors({ ...errors });
    return !errorsExits;
  };

  const handleEmailChange = (
    event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    setEmail(event.target.value);
  };

  const handlePasswordChange = (
    event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    setPassword(event.target.value);
  };

  const handleLogIn = () => {
    if (!validateData()) {
      return;
    }
    setLoginLoading(true);
    axios
      .post(
        signInEndPoint,
        { email: email, password: password, returnSecureToken: true },
        {
          headers: {
            "content-type": "application/json",
          },
        }
      )
      .then((res) => {
        console.log("Sign In response is: ", res);
        localStorage.setItem("refresh", res.data.refreshToken);
        axiosInstance
          .get("/readUser", { params: { userID: res.data.localId } })
          .then((uData) => {
            setLoginLoading(false);
            const loginTime = new Date();
            const lastLogin = parseInt(format(loginTime, "T"));
            dispatch(
              userLoggedIn({
                ...uData.data.content,
                idToken: res.data.idToken,
                lastLogin,
              })
            );
            navigate("/boardsView");
          })
          .catch((userError) => {
            console.log("error while getting user info: ", userError);
            setLoginLoading(false);
          });
      })
      .catch((err) => {
        if (err.response.data.error.message == "EMAIL_NOT_FOUND") {
          errors.email = "Email not found.";
        } else if (err.response.data.error.message == "INVALID_PASSWORD") {
          errors.password = "Password incorrect.";
        }
        setLoginLoading(false);

        setErrors({ ...errors });
        console.log("Failed to sign user in: ", err);
      });
  };

  const handleSignUpCLick = () => {
    navigate("/signup");
  };

  return (
    <div className="login-container">
      <Grid className="login" direction="row" container spacing={2}>
        <Grid
          className="login-left-container"
          justifyContent="center"
          alignItems="center"
          container
          item
          md={5}
          lg={5}
          xl={5}
        >
          <Grid
            className="login-inner-grid"
            justifyContent="center"
            alignItems="center"
            container
            item
            xs={12}
          >
            <Grid
              className="login-inner-grid"
              container
              item
              direction="column"
              justifyContent="center"
              alignItems="center"
              xs={12}
            >
              <img className="login-logo-image" src={Logo} />
              <h1 className="login-app-name">Bulletin</h1>
              <h3 className="login-app-description">
                Organize your household in a smart way{" "}
              </h3>
            </Grid>
            <Grid
              className="login-inner-grid"
              container
              item
              direction="column"
              justifyContent="center"
              alignItems="center"
              xs={12}
            >
              <h4 className="login-signup-redirect-text">New Here ?</h4>
              <Button
                onClick={handleSignUpCLick}
                className="login-signup-redirect-button"
                variant="contained"
              >
                Sign up now!
              </Button>
            </Grid>
          </Grid>
        </Grid>
        <Grid
          className="login-inner-grid"
          container
          justifyContent="center"
          alignItems="center"
          item
          xs={12}
          md={7}
          lg={7}
          xl={7}
          direction="column"
          spacing={4}
        >
          <Grid item>
            <h1 className="login-title">Log In</h1>
          </Grid>
          <Grid className="login-inner-grid-item" item>
            <p>Email</p>
            <TextField
              id="login-username-email-field"
              className="login-text-field"
              InputLabelProps={{ shrink: false }}
              value={email}
              onChange={handleEmailChange}
              label={email === "" ? "Email" : ""}
              variant="outlined"
              error={errors.email != ""}
              helperText={errors.email}
            />
          </Grid>
          <Grid className="login-inner-grid-item" item>
            <p>Password</p>
            <TextField
              id="login-password-field"
              className="login-text-field"
              InputLabelProps={{ shrink: false }}
              value={password}
              onChange={handlePasswordChange}
              label={password === "" ? "Password" : ""}
              variant="outlined"
              type="password"
              error={errors.password != ""}
              helperText={errors.password}
            />
          </Grid>
          <Grid item>
            <SpinnerButton
              onClick={handleLogIn}
              className="login-button"
              title="Log In"
              loading={loginLoading}
            />
          </Grid>
        </Grid>
      </Grid>
    </div>
  );
};

export default LogIn;
