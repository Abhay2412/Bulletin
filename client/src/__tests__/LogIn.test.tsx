import React from "react";
import LogIn from "../pages/LogIn/LogIn";
import { configure, mount, ReactWrapper } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

configure({ adapter: new Adapter() });

describe("Testing <LogIn /> Component", () => {
  it("should change email field value", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <LogIn />
        </BrowserRouter>
      </Provider>
    );
    wrapper
      .find('input[type="text"]')
      .find("#login-username-email-field")
      .simulate("change", {
        target: { value: "test@domain.com" },
      });
    expect(
      wrapper
        .find('input[type="text"]')
        .find("#login-username-email-field")
        .get(0).props.value
    ).toEqual("test@domain.com");
  });

  it("should chnage password field value", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <LogIn />
        </BrowserRouter>
      </Provider>
    );
    wrapper
      .find('input[type="password"]')
      .find("#login-password-field")
      .simulate("change", {
        target: { value: "12345678" },
      });
    expect(
      wrapper
        .find('input[type="password"]')
        .find("#login-password-field")
        .get(0).props.value
    ).toEqual("12345678");
  });

  it("should have email error", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <LogIn />
        </BrowserRouter>
      </Provider>
    );
    wrapper
      .find('input[type="password"]')
      .find("#login-password-field")
      .simulate("change", {
        target: { value: "12345678" },
      });

    wrapper.find("button").find(".login-button").simulate("click");
    expect(
      wrapper.find("p").find("#login-username-email-field-helper-text").text()
    ).toEqual("Please enter an email.");
  });

  it("should have password error", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <LogIn />
        </BrowserRouter>
      </Provider>
    );
    wrapper
      .find('input[type="text"]')
      .find("#login-username-email-field")
      .simulate("change", {
        target: { value: "test@domain.com" },
      });

    wrapper.find("button").find(".login-button").simulate("click");
    expect(
      wrapper.find("p").find("#login-password-field-helper-text").text()
    ).toEqual("Please enter a password.");
  });
});
