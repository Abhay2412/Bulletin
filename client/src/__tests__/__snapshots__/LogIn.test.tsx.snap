// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <LogIn /> Component should render properly 1`] = `
<div
  className="login-container"
>
  <div
    className="MuiGrid-root login MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root login-left-container MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center MuiGrid-grid-md-5 MuiGrid-grid-lg-5 MuiGrid-grid-xl-5"
    >
      <div
        className="MuiGrid-root login-inner-grid MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12"
      >
        <div
          className="MuiGrid-root login-inner-grid MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12"
        >
          <img
            className="login-logo-image"
            src="logo.svg"
          />
          <h1
            className="login-app-name"
          >
            Bulletin
          </h1>
          <h3
            className="login-app-description"
          >
            Organize your household in a smart way
             
          </h3>
        </div>
        <div
          className="MuiGrid-root login-inner-grid MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12"
        >
          <h4
            className="login-signup-redirect-text"
          >
            New Here ?
          </h4>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained login-signup-redirect-button"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Sign up now!
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root login-inner-grid MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12 MuiGrid-grid-md-7 MuiGrid-grid-lg-7 MuiGrid-grid-xl-7"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <h1
          className="login-title"
        >
          Log In
        </h1>
      </div>
      <div
        className="MuiGrid-root login-inner-grid-item MuiGrid-item"
      >
        <p>
          Email
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root login-text-field"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink={false}
            htmlFor="login-username-email-field"
            id="login-username-email-field-label"
          >
            Email
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="login-username-email-field"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Email
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-root login-inner-grid-item MuiGrid-item"
      >
        <p>
          Password
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root login-text-field"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink={false}
            htmlFor="login-password-field"
            id="login-password-field-label"
          >
            Password
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="login-password-field"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root login-button css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Log In
        </button>
      </div>
    </div>
  </div>
</div>
`;
